### DISCUSSION TITLE
Best way to restrict references?

### DISCUSSION THREAD

**Discussion by paulfrench** [eclipse-langium/langium#360] [Q&A] [answered]
Here is a grammar file I've simplified to demonstrate my problem...
```
grammar Cleese
hidden(WS, SL_COMMENT, ML_COMMENT)

Model:
	(elements+=Element)*;

Element:
	Entity;

Entity:
	'entity' name=ID ('extends' superType = [Entity])? '{'
		(properties+=Property)*
    ('keys' ':' '[' keys+=[Property]("," keys+=[Property])* ']' )?
	'}';

Property:
	name=ID ':' type=AllTypes;

AllTypes:
  Type | 'List' '<' Type '>';

Type:
	EntityReference | SimpleDataType;

EntityReference:
	type=[Entity];

SimpleDataType:
	type=PrimitiveType;

PrimitiveType returns string:
	"number" | "string" | "date" | "boolean";


terminal WS: /\s+/;
terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
terminal SL_COMMENT: /\/\/[^\n\r]*/;

terminal ID: /[_a-zA-Z][\w_]*/;
```
I can create a Entity no problem and auo-completion works great for the supertype and for the keys.
```
entity Address {
  id: number
  line1: string
  postcode: string
  keys : [ id ]
}

entity Person {
  id: number
  firstname: string
  surname: string
  address: Address
  keys : [ id ]
}
```
The question is how to I modify the grammar so that the auto-suggestions for the keys that can be used in the Person entity is only the properties that are primitive types? At the moment, "id", "firstname", "surname" and "address" get auto suggested, but a key cannot be a complex object and so I do not want the "address" to be auto-suggested as a key?

I'm missing something somewhere in my understanding of how to construct the grammar and possibly not try and do everything in the grammar itself and instead write validation rules? Not sure, any guidance would be great!

**Comment by spoenemann** [comment, marked as answer]
There are multiple ways to solve that:

 * Split the `Property` rule in two variants, one for primitive types and one for complex types, and use the primitive variant in the cross-reference (not recommended).
 * Change the `ScopeProvider` service so it filters the properties to include only those with primitive types (not recommended).
 * Add a validation check that shows an error when a property with complex type is used as key, and adapt the `CompletionProvider` service to show only the properties with primitive type for that cross-reference.

I recommend the third approach because it results in a much better error message to the users: Instead of "Could not resolve reference..." they will see something like "Only primitive types can be used as keys".

**Reply by paulfrench** [reply]
Many thanks @spoenemann that has given me some direction and will save me considerable time!

**Reply by paulfrench** [reply]
Hi @spoenemann I don't see a `CompletionProvider` service, can you elaborate? Cheers

**Reply by msujew** [reply]
@paulfrench The service can be accessed using `LangiumServices.lsp.completion.CompletionProvider`:

https://github.com/langium/langium/blob/27d38066fb8b79e267be94d2027d2da88322b2d7/packages/langium/src/services.ts#L60

Its interface and default implementation can be found here:

https://github.com/langium/langium/blob/main/packages/langium/src/lsp/completion/completion-provider.ts

You should probably override this method:

https://github.com/langium/langium/blob/27d38066fb8b79e267be94d2027d2da88322b2d7/packages/langium/src/lsp/completion/completion-provider.ts#L124-L137

@spoenemann We should consider adding a `filterCrossReferences` method to it as done in Xtext.
